### ISSUE TITLE
Fully coupling matrix is set as default  if a pure Newton is chosen 

### ISSUE THREAD

**Issue by fdkong** [idaholab/moose#9968]
## Rationale
Looks like users usually forget to use a full coupling matrix when they are trying to use pure Newton. Newton depends on the accuracy of Jacobian.  If some entires are ignored automatically, a bad convergence may occur. 

## Description
In `FEProblemBase`, `_coupling = Moose::COUPLING_FULL` if Newton is used. 

## Impact
Affect users who are using the pure Newton.

**Comment by mangerij** [comment]
I support this. Its not clear why full = true wasnt the default to begin
with :)

J

**Comment by friedmud** [comment]
Because MOOSE is intended for MULTI-physics problems.  Like solving 10s to thousands of equations where `full = true` is a terrible idea... even for `Newton`.

Why should we assume that all physics are coupled to eachother and waste memory and time?

Even in quasi-static thermomechanics `full = true` is often not right (the displacements might not contribute back to the temperature Jacobian).... and that's a _really_ simple problem.

I am *absolutely* against this change.  If you want to use `full = true`... go ahead and do it... but know that you are likely wasting memory time (but at least you're doing it explicity).

**Comment by fdkong** [comment]
I agree that ` full=true` takes more memory. But I also concern on the converge rate. Newton highly depends on the accuracy of Jacobian, and it may diverge if we just throw away off-diagonal entries silently. Then some users just complain that my code with XXX package converges quadratically, but it diverges within MOOSE.    

Explicitly changing to `full` might not be a good idea. Could we do something to let users know that they should pay attention to Jacobian when they are using a pure Newton?

**Comment by lindsayad** [comment]
It seems to me like this is an argument of whether we should default to integrity of a solution method or default to efficient resource utilization. I would be inclined to default to integrity of the solution method; e.g. I think the proposed change is a good one. `NEWTON` generally just won't work unless you have the full and correct Jacobian. As a user, I would lean towards having something work slowly than to not work at all.

**Comment by friedmud** [comment]
But you also get the people who are saying: "MOOSE uses too much memory!" when they've specified full=true...

How about this: what if we add a new `solve_type` option: `FULL_NEWTON`.  This would be `Newton` with an implied `full = true`.  It can come complete with a big warning saying that you are most likely wasting memory and should transition over to using `NEWTON` with more specific Preconditioner options in the future...

**Comment by YaqiWang** [comment]
Actually, MOOSE can be smarter on this. MOOSE can check all the kernels to see if there is any offDiagJacobian between any two variables, if there is any, MOOSE turns the coupling to true when solve type is NEWTON, otherwise to false.

**Comment by fdkong** [comment]
If `FULL_NEWTON` is the right option for some particular problems, but we are still throwing out a big warning?  I am not sure if it is a good way.  I may prefer to have `FULL_NEWTON` without warning.

**Comment by friedmud** [comment]
@YaqiWang I was thinking about this as well... and it could be possible.

In order to provide off-diagonal coupling you MIUST call `coupled()` (to get the unique identifier)... so we really could keep track of what is coupled to what that way and turn on just the right blocks by default with NEWTON.

The next question would be: what should be the default with PJFNK?  I would still vote for the block-diagonal... but now there will be a pretty severe inconsistency with NEWTON...

**Comment by permcody** [comment]
I like full for newton and block-diagonal for PJFNK. We could add another information line in the system information printout about what's going on with the preconditioning/Jacobian. Sure a lot of people don't bother to look at this, but I would say it's their own fault for ignoring that kind of information.

**Comment by andrsd** [comment]
> The next question would be: what should be the default with PJFNK? I would still vote for the block-diagonal... but now there will be a pretty severe inconsistency with NEWTON...

I think it is ok that different `solve_type`s have different defaults. When one selects PJFNK, they will soon introduce `[Preconditioning]` block and will start changing the defaults. So, I think having block-diagonal preconditioning as default is a good starting point. And I like @YaqiWang's suggestion that for NEWTON, we should figure out the coupling matrix and use it.

**Comment by friedmud** [comment]
I concur with @andrsd
